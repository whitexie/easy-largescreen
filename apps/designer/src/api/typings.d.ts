declare namespace API {
  interface CreateDatasetDto {
    /** 数据集名称 */
    name: string
    /** 数据集描述 */
    description?: string
    /** 数据集code */
    datasetCode: string
    /** 数据集字段 */
    fields: CreateFieldDto[]
  }

  interface CreateFieldDto {
    /** 所属表 */
    datasetCode?: string
    /** 字段code */
    fieldCode: string
    /** 字段名称 */
    name: string
    /** 字段描述 */
    description: string
    /** 字段类型 */
    valueType: 'string' | 'number' | 'date'
  }

  interface Dataset {
    /** 数据Id */
    id: string
    /** 数据创建时间 */
    created_time: string
    /** 数据更新时间 */
    updated_time: string
    /** 数据集名称 */
    name: string
    /** 数据集对应的数据表 */
    datasetCode: string
    /** 数据集描述 */
    description?: string
    /** 数据集字段 */
    fields: Field[]
  }

  interface Field {
    /** 数据Id */
    id: string
    /** 数据创建时间 */
    created_time: string
    /** 数据更新时间 */
    updated_time: string
    /** 所属表 */
    datasetCode: string
    /** 字段code */
    fieldCode: string
    /** 字段名称 */
    name: string
    /** 字段描述 */
    description: string
    /** 字段值类型 */
    valueType: 'string' | 'number' | 'date'
    /** 所属数据集ID */
    datasetId: string
  }

  interface FieldDto {
    fieldCode: string
    id: string
    name: string
    valueType: string
    fieldType: string
  }

  interface GetDatasetDetailByIdParams {
    datasetId: string
  }

  interface GetFieldsByDatasetIdParams {
    datasetId: string
  }

  interface MetricFieldDto {
    fieldCode: string
    id: string
    name: string
    valueType: string
    fieldType: string
    calculateType: string
  }

  interface Object {}

  interface ResponseWrapper {
    data: Record<string, any>
    error: number
    msg: string
  }

  interface SearchDataDto {
    datasetId: string
    dimensionFields: FieldDto[]
    metricFields: MetricFieldDto[]
  }
}
