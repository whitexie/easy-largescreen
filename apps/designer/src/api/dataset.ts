/**
 * 该文件为 @umijs/openapi 插件自动生成，请勿随意修改。如需修改请通过配置 openapi.config.ts 进行定制化。
 */

import { request } from '@/utils';

/** 创建数据集 POST /dataset/createDataset */
export async function createDataset(body: API.CreateDatasetDto, options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: API.Dataset }>(`/api/dataset/createDataset`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** 获取数据集详情, 含字段结构 GET /dataset/getDatasetDetailById */
export async function getDatasetDetailById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.GetDatasetDetailByIdParams,
  options?: Record<string, any>,
) {
  return request<API.ResponseWrapper & { data?: API.Dataset }>(
    `/api/dataset/getDatasetDetailById`,
    {
      method: 'GET',
      params: {
        ...params,
      },
      ...(options || {}),
    },
  );
}

/** 获取数据集列表 GET /dataset/getDatasetList */
export async function getDatasetList(options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: API.Dataset[] }>(`/api/dataset/getDatasetList`, {
    method: 'GET',
    ...(options || {}),
  });
}

/** 获取数据集字段结构 GET /dataset/getFieldsByDatasetId */
export async function getFieldsByDatasetId(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.GetFieldsByDatasetIdParams,
  options?: Record<string, any>,
) {
  return request<API.ResponseWrapper & { data?: API.Field[] }>(
    `/api/dataset/getFieldsByDatasetId`,
    {
      method: 'GET',
      params: {
        ...params,
      },
      ...(options || {}),
    },
  );
}

/** 获取数据集数据 POST /dataset/searchData */
export async function searchData(body: API.SearchDataDto, options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: Record<string, string | number>[] }>(`/api/dataset/searchData`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}
