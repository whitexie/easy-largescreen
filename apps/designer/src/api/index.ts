export * from './dataset';
export * from './mock-end-service';
