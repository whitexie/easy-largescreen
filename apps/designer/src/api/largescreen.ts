/**
 * 该文件为 @umijs/openapi 插件自动生成，请勿随意修改。如需修改请通过配置 openapi.config.ts 进行定制化。
 * */

import { request } from '@/utils';

/** 创建大屏 POST /largescreen/create */
export async function create(body: API.SaveLargescreenDto, options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: string }>(`/api/largescreen/create`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** 大屏详情 GET /largescreen/detail/${param0} */
export async function detail(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.DetailParams,
  options?: Record<string, any>,
) {
  const { id: param0, ...queryParams } = params;
  return request<API.ResponseWrapper & { data?: API.LargeScreenDetailDto }>(
    `/api/largescreen/detail/${param0}`,
    {
      method: 'GET',
      params: { ...queryParams },
      ...(options || {}),
    },
  );
}

/** 大屏列表 GET /largescreen/list */
export async function list(options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: API.LargeScreenDto[] }>(`/api/largescreen/list`, {
    method: 'GET',
    ...(options || {}),
  });
}

/** 发布大屏 POST /largescreen/release */
export async function release(body: API.ReleaseLargescreenDto, options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: string }>(`/api/largescreen/release`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** 删除大屏 DELETE /largescreen/remove */
export async function remove(body: API.DeleteLargescreenDto, options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: string }>(`/api/largescreen/remove`, {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}
