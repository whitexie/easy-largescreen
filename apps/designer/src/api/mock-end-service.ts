import type { Field } from '@/types/charts';
import localforage from 'localforage';

const dataTableStore = localforage.createInstance({ name: 'datasetData' });

interface AnalysisDataParams {
  datasetId: string
  dimensionFields: Field[]
  metricFields: Field[]
}

function transformData(data: any[], fields: Field[]): Record<string, any>[] {
  return data.reduce((acc, cur) => {
    const row: Record<string, any> = {};
    fields.forEach((field) => {
      row[field.id] = cur[field.fieldCode];
    });
    acc.push(row);
    return acc;
  }, []);
}

export async function getAnlysisData(params: AnalysisDataParams): Promise<Record<string, any>[]> {
  const { datasetId, dimensionFields, metricFields } = params;
  let data = await dataTableStore.getItem<any[]>(datasetId);
  if (!data) {
    return [];
  }

  data = transformData(data, dimensionFields.concat(metricFields));

  return data;
}
