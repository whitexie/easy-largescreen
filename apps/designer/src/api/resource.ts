/**
 * 该文件为 @umijs/openapi 插件自动生成，请勿随意修改。如需修改请通过配置 openapi.config.ts 进行定制化。
 */

import { request } from '@/utils';

/** 获取上传凭证 POST /resource/getUploadToken */
export async function getUploadToken(options?: Record<string, any>) {
  return request<API.ResponseWrapper & { data?: API.UploadTokenDto }>(
    `/api/resource/getUploadToken`,
    {
      method: 'POST',
      ...(options || {}),
    },
  );
}
