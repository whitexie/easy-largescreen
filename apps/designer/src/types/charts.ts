export interface FieldConfigItem {
  name: string
  type: 'dimension' | 'metric'
}
