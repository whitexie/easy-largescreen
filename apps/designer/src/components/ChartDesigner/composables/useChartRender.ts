import type { BoxId, ChartRederStateOptions, ChartRenderState, DropBoxSettings, Field, FieldType, OriginalField } from '@/types/charts';
import api from '@/api';
import { CHART_MAPPING } from '@/components/Charts/constants/chart';
import { CalculateType } from '@/types/charts';
import { generateId, hasOwn } from '@yss/utils';
import { pick } from 'lodash-es';
import { computed, reactive, unref, type WatchHandle } from 'vue';

const CHARTS = Object.values(CHART_MAPPING);

export function useChartRender(options?: Partial<ChartRederStateOptions>) {
  const state = reactive<ChartRenderState>({
    datasetId: '',
    chartType: CHARTS[0].id,
    dropBoxSettings: {
      xAxis: {
        id: 'xAxis',
        title: 'X轴',
        fieldType: 'dimension',
        fields: [],
      },
      yAxis: {
        id: 'yAxis',
        title: 'Y轴',
        fieldType: 'metric',
        fields: [],
      },
    },
    chartProps: {},
  });
  const data = reactive<any[]>([]);

  const watchHandle: WatchHandle | null = null;

  const datasetId = computed({
    get() {
      return state.datasetId;
    },
    set(val) {
      state.datasetId = val;
    },
  });

  const chartConfig = computed(() => {
    return unref(state);
  });

  const listenSource = computed(() => {
    return Object.values(state.dropBoxSettings);
  });

  function getChartConfig() {
    return structuredClone(unref(state));
  }

  function updateState(options: Partial<ChartRederStateOptions>) {
    if (watchHandle) {
      watchHandle();
    }
    const keys = Object.keys(options) as Array<keyof ChartRederStateOptions>;
    keys.forEach((key) => {
      if (hasOwn(state, key) && options[key] !== undefined) {
        (state[key] as any) = options[key];
      }
    });

    if (options.chartType && !hasOwn(options, 'dropBoxSettings')) {
      if (options.chartType in CHART_MAPPING) {
        const chartConfig = CHART_MAPPING[options.chartType];
        const dropBoxSettings = structuredClone(chartConfig.dropBoxSettings);
        for (const key in dropBoxSettings) {
          dropBoxSettings[key].fields = [];
        }
        state.dropBoxSettings = dropBoxSettings as Record<string, Required<DropBoxSettings>>;
        state.chartProps = structuredClone(chartConfig.props);
      }
    }
  }

  function createField(fieldOption: OriginalField, type: FieldType): Field {
    const whiteList: (keyof OriginalField)[] = ['fieldCode', 'name', 'valueType'];
    return {
      ...pick(fieldOption, whiteList),
      fieldType: type,
      datasetId: state.datasetId,
      id: generateId(),
      calculateType: CalculateType.SUM,
    };
  }

  function addField(boxId: BoxId, fieldOption: OriginalField, index?: number) {
    const box = state.dropBoxSettings[boxId];
    const length = box.fields.length;
    const fieldType = box.fieldType;
    const field = createField(fieldOption, fieldType);
    if (field.valueType !== 'number') {
      field.calculateType = CalculateType.COUNT;
    }
    state.dropBoxSettings[boxId].fields.splice(index || length, 0, field);

    return field;
  }

  function removeField(boxId: BoxId, fieldId: string) {
    const box = state.dropBoxSettings[boxId];
    const index = box.fields.findIndex(field => field.id === fieldId);
    if (index !== -1) {
      box.fields.splice(index, 1);
    }
  }

  function conversionFields() {
    const dimensionFields: Field[] = [];
    const metricFields: Field[] = [];

    Object.values(state.dropBoxSettings).forEach((box) => {
      box.fields.forEach((field) => {
        if (field.fieldType === 'dimension') {
          dimensionFields.push(unref(field));
        }
        else {
          metricFields.push(unref(field));
        }
      });
    });

    return {
      dimensionFields,
      metricFields,
    };
  }

  async function requestData() {
    const { dimensionFields, metricFields } = conversionFields();
    const { datasetId } = state;
    const result = await api.dataset.searchData({
      datasetId,
      dimensionFields,
      metricFields,
    });
    data.splice(0, data.length);
    data.push(...result.data);

    return result.data;
  }

  if (options) {
    updateState(options);
  }

  watch(
    () => listenSource.value,
    () => {
      // console.log('listenSource changed');
      requestData();
    },
    {
      deep: true,
    },
  );

  return {
    state,
    datasetId,
    data,
    chartConfig,
    addField,
    removeField,
    requestData,
    getChartConfig,
  };
}
