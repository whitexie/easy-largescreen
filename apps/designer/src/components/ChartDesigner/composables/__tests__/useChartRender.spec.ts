import api from '@/api';
import { CalculateType, type ChartRederStateOptions, type OriginalField } from '@/types/charts';
import * as utils from '@yss/utils';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { nextTick } from 'vue';
import { useChartRender } from '../useChartRender';

describe('useChartRender', () => {
  it('应该正确初始化状态', () => {
    const { state } = useChartRender();
    expect(state.datasetId).toBe('');
    expect(state.dropBoxSettings).toHaveProperty('xAxis');
    expect(state.dropBoxSettings).toHaveProperty('yAxis');
  });

  describe('应该正确初始化dropBoxSettings', () => {
    it('chartRederStateOptions 包含 dropBoxSettings时', () => {
      const settings: ChartRederStateOptions['dropBoxSettings'] = {
        xAxis: {
          id: 'xAxis',
          title: 'X轴',
          fieldType: 'dimension',
          fields: [{
            name: '日期',
            valueType: 'date',
            fieldCode: 'ydqlmbGj',
            fieldType: 'dimension',
            datasetId: 'id-1726113264669dSQYOyWg',
            id: 'TU2MS9S4',
            calculateType: CalculateType.COUNT,
          }],
        },
        yAxis: {
          id: 'yAxis',
          title: 'Y轴',
          fieldType: 'metric',
          fields: [{
            name: '销售金额',
            valueType: 'number',
            fieldCode: 'yAQJDYrC',
            fieldType: 'metric',
            datasetId: 'id-1726113264669dSQYOyWg',
            id: '9PB11S17',
            calculateType: CalculateType.SUM,
          }],
        },
      };

      const options: Partial<ChartRederStateOptions> = {
        dropBoxSettings: settings,
      };

      const { state } = useChartRender(options);
      expect(state.dropBoxSettings).toEqual(settings);
    });

    it('chartRederStateOptions 只有chartType 时', () => {
      const settings: ChartRederStateOptions = {
        chartType: 'map',
      };

      const { state } = useChartRender(settings);

      expect(state.dropBoxSettings?.coordinates).toMatchSnapshot();
    });
  });

  it('应该正确更新datasetId', async () => {
    const { datasetId } = useChartRender();
    datasetId.value = 'test-dataset';
    await nextTick();
    expect(datasetId.value).toBe('test-dataset');
  });

  it('应该正确添加度量字段', () => {
    vi.spyOn(utils, 'generateId')
      .mockReturnValueOnce('test-field');
    const { state, addField } = useChartRender();
    const fieldOption = {
      id: 'test-field',
      fieldCode: 'amount',
      name: 'Test Field',
      valueType: 'number',
    } as OriginalField;

    addField('yAxis', fieldOption, 0);

    expect(state.dropBoxSettings.yAxis.fields).toHaveLength(1);
    expect(state.dropBoxSettings.yAxis.fields[0]).toMatchObject({
      fieldCode: fieldOption.fieldCode,
      name: fieldOption.name,
      valueType: fieldOption.valueType,
      fieldType: 'metric',
      calculateType: CalculateType.SUM,
    });
  });

  it('应该正确添加维度字段', () => {
    vi.spyOn(utils, 'generateId')
      .mockReturnValueOnce('provinces-1');

    const { state, addField } = useChartRender();
    const fieldOption = {
      fieldCode: 'test_field',
      name: 'Test Field',
      valueType: 'string',
    } as OriginalField;

    addField('xAxis', fieldOption);

    expect(state.dropBoxSettings.xAxis.fields).toHaveLength(1);
    expect(state.dropBoxSettings.xAxis.fields[0]).toMatchObject({
      id: 'provinces-1',
      fieldCode: fieldOption.fieldCode,
      name: fieldOption.name,
      valueType: fieldOption.valueType,
      fieldType: 'dimension',
      calculateType: CalculateType.COUNT,
    });

    vi.clearAllMocks();
  });

  it('应该正确删除字段', () => {
    const { state, addField, removeField } = useChartRender();
    const fieldOption = {
      id: 'test-field',
      name: 'Test Field',
      valueType: 'number',
    } as OriginalField;

    const field = addField('yAxis', fieldOption, 0);
    removeField('yAxis', field.id);

    expect(state.dropBoxSettings.yAxis.fields).toHaveLength(0);
  });

  it('应该使用提供的选项初始化状态', () => {
    const options = {
      datasetId: 'initial-dataset',
    };
    const { state } = useChartRender(options);
    expect(state.datasetId).toBe('initial-dataset');
  });

  describe('requestData', () => {
    beforeEach(() => {
      vi.spyOn(utils, 'generateId')
        .mockReturnValueOnce('provinces-1')
        .mockReturnValueOnce('amount-1');
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });
    it('应正确请求数据', async () => {
      const { addField, requestData, datasetId } = useChartRender();

      datasetId.value = 'test-dataset';

      addField('xAxis', {
        id: 'provinces',
        fieldCode: 'provinces',
        name: '省份',
        valueType: 'string',
      });

      addField('yAxis', {
        id: 'amount',
        name: '金额',
        fieldCode: 'amount',
        valueType: 'number',
      });

      const expecteData = [
        { 'provinces-1': '广东', 'amount-1': 100 },
        { 'provinces-1': '广西', 'amount-1': 200 },
      ];

      vi.spyOn(api.dataset, 'searchData')
        .mockReturnValueOnce(Promise.resolve({ error: 0, data: expecteData, msg: 'success' }));

      const data = await requestData();

      expect(data).toEqual(expecteData);
    });
  });
});
