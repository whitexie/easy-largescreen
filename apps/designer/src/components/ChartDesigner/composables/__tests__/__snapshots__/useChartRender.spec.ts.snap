// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useChartRender > 应该正确初始化dropBoxSettings > chartRederStateOptions 只有chartType 时 1`] = `
{
  "fieldType": "dimension",
  "fields": [],
  "id": "coordinates",
  "title": "坐标",
}
`;
